### ISSUE TITLE
Benchmark Issues

### ISSUE THREAD

**Issue by Brandon-T** [bddicken/languages#71]
1. These benchmarks aren't just testing the 1 billion loops, they test how long it takes for the OS to launch the program, and how long it takes to print to the console.
2. It also tests how long it takes the JVM to start up as well, since it does `run java java.code 40` where java will have to start a JVM instance, and then run the code. Likewise for other interpreted code such as python.
3. They are testing print statements like `printf` which does I/O to `$CONOUT`/`stdout` for example. Each language will print differently. 
4. `rand` in C will outperform Java's `Integer.random` by a lot due to C's rand not being thread-safe vs. Java's. `rand_r` would be equivalent in C, or `ThreadLocalRandom.current().nextInt()` in Java.
5. The benchmarks aren't testing the fastest possible way to do things in each language. In C, you would not initialize the array to 0 and then overwrite each index. You'd skip initialization and just write to the array.

Overall, we should be testing only the 1 billion loops by setting up a timer in each language, looping, then taking the difference from start to finish, rather than launching the programs as-is.

A cold start JVM will take quite a bit of time!

C should be using `-O3 -funroll-loops -march=native -ftree-vectorize -ffast-math`

**Comment by nervenes** [comment]
Yeah.. let's solve a useless benchmark by making it more useless 👍

Seriously though, these types of benchmarks doesn't benchmark anything, especially when they're compiled - like the swift vs rust vs c vs c++ will just be llvm vs llvm doing no iterations as they're optimized away. and they'll always be faster than a language that'll require a runtime or can't optimize away the loops, which again - resembles barely anything meaningful in the real world.

**Comment by Brandon-T** [comment]
@nervenes By your logic, we shouldn't ever benchmark anything at all. The benchmarks test compiler optimizations and instruction generation as do all benchmarks. That's the point. C was created to be as close as possible to bare metal.
They'd only be doing no iterations IF the compiler can optimize it away if it finds the variable unused.

> they'll always be faster than a language that'll require a runtime or can't optimize away the loops, which again

That's a problem with that language/compiler/runtime then.

**Comment by nervenes** [comment]
No, what you're suggesting is the opposite. To actually benchmark the runtime you'd need a long running process actually computing something, of which is neither what you're suggesting nor what the current benchmarks are.

**Comment by Brandon-T** [comment]
> No, what you're suggesting is the opposite. To actually benchmark the runtime you'd need a long running process actually computing something, of which is neither what you're suggesting nor what the current benchmarks are.

I did suggest that if you read the post properly. It says NOT to test launching the JVM or python interpreter as part of the benchmark, and to use a timer + the benchmark code to test the actual code execution.

**Comment by nervenes** [comment]
You appear to not understand how this works, at all.

**Comment by Brandon-T** [comment]
You appear not to be able to comprehend what I wrote, and is off on a tangent.
I'm aware the benchmark is not sorting an array of 1b integers testing a specific algorithm like quicksort or bubble or w/e "real world" test you want it to do.
So yes, it does not have a real world impact and you can't post on your blog that X language's implementation of Y is faster.

But the point of benchmarks is a: A compiler test. b: Instruction generation. c: Runtime. d: Optimizations.
That applies here. The compiled code for all the languages WILL be faster than the interpreted ones. Everyone knows this.

What you're missing, is that not all compiled code is created equal. Not all optimizations are done equally. Not all language rules allow certain optimizations.

So yes, while it's LLVM vs. LLVM for different languages when you get down to it, it shows the implementations and optimizations that can or can't be applied to certain languages or runtimes.
We do comparisons with clang vs. gcc all the time, for meaningless code just to see what the performance is like or code generation or optimizations, etc.

You could always take it up with the author if you wanted something "meaningful". 
Maybe open your own issue telling them you'd like to see something meaningful in the real world instead of a benchmark that just sums integers that could be unrolled or completely optimized away.

**Comment by nervenes** [comment]
> TLDR: My point was simply to point out that Java and other runtimes aren't technically treated fairly in THIS benchmark.

I'm not disagreeing with this fact. I'm disagreeing with your solution to it.

> we should be testing only the 1 billion loops by setting up a timer in each language, looping, then taking the difference from start to finish

This still doesn't solve the fact that the compiled languages will likely optimize away the entire loop while others don't, and what you fail to understand is that you're no longer comparing **equal work**, and you also seem to confuse that with **equal codegen**, which is not the point. In order to see if one language is faster than the other, you need to do similar work, during a longer period. The current benchmarks, and the one you're arguing to implement, are flawed for the intent of comparing language performance.

**Comment by cyrusmsk** [comment]
> Yeah.. let's solve an already useless benchmark by making it more useless 👍
> 
> Seriously though, these types of benchmarks doesn't benchmark anything, especially when they're compiled - like the swift vs rust vs c vs c++ will just be llvm vs llvm doing no iterations as they're optimized away. and they'll always be faster than a language that'll require a runtime or can't optimize away the loops, which again - resembles barely anything meaningful in the real world.

Of course they are not. They depend on many things.
For example here (https://github.com/jinyus/related_post_gen) Java outperform Swift and other LLVM-based solutions.
The issue of calculating time only for real work and not for the I/O, VM start and other things is relevant.
And this is what was reasonably mentioned by the issue author.

For real usage scenario you probably gonna run environment only once, and then in the cycle process requests/data many times.

**Comment by cyrusmsk** [comment]
> No, what you're suggesting is the opposite. To actually benchmark the runtime (i.e code execution as you're referring it to) you'd need a long running process actually computing something, of which is neither what you're suggesting nor what the current benchmarks are.

But I agree that the calculation in the current form is not very "fair", because some advanced optimizations could just squash internal loop, while others can't.
That's why I propose to make calculations more tricky, so LLVM can't skip it.

**Comment by Brandon-T** [comment]
@nervenes Nothing I suggested makes it no longer compare **equal work**.

> This still doesn't solve the fact that the compiled languages will likely optimize away the entire loop while others don't,

You state this is fact but it doesn't and it takes a few seconds on godbolt just to verify that: https://godbolt.org/z/j3GWehxed. 
The C code generates the following Assembly with the loops intact with:

`-O3`
```c
.LC0:
        .string "%d\n"
main:
        push    rbx
        mov     edx, 10
        sub     rsp, 40000
        mov     rdi, QWORD PTR [rsi+8]
        xor     esi, esi
        call    strtol
        mov     ebx, eax
        call    rand
        mov     rdi, rsp
        movsx   r8, eax
        cdq
        imul    r8, r8, 1759218605
        sar     r8, 44
        sub     r8d, edx
        imul    edx, r8d, 10000
        sub     eax, edx
        movsx   r8, eax
.L2:
        mov     esi, DWORD PTR [rdi]
        xor     ecx, ecx
.L3:
        mov     eax, ecx
        add     ecx, 1
        cdq
        idiv    ebx
        add     esi, edx
        cmp     ecx, 100000
        jne     .L3
        add     esi, r8d
        add     rdi, 4
        lea     rax, [rsp+40000]
        mov     DWORD PTR [rdi-4], esi
        cmp     rdi, rax
        jne     .L2
        mov     esi, DWORD PTR [rsp+r8*4]
        mov     edi, OFFSET FLAT:.LC0
        xor     eax, eax
        call    printf
        add     rsp, 40000
        xor     eax, eax
        pop     rbx
        ret
```

`-O3 -funroll-loops -march=native -ftree-vectorize -ffast-math`:
```c
.LC0:
        .string "%d\n"
main:
        push    rbp
        mov     edx, 10
        mov     rbp, rsp
        push    rbx
        and     rsp, -32
        sub     rsp, 40000
        mov     rdi, QWORD PTR [rsi+8]
        xor     esi, esi
        call    strtol
        mov     ebx, eax
        call    rand
        mov     rsi, rsp
        mov     edi, eax
        cdqe
        imul    rax, rax, 1759218605
        mov     edx, edi
        sar     edx, 31
        sar     rax, 44
        sub     eax, edx
        imul    ecx, eax, 10000
        sub     edi, ecx
.L2:
        mov     r9d, DWORD PTR [rsi]
        xor     r8d, r8d
.L3:
        mov     eax, r8d
        cdq
        idiv    ebx
        lea     eax, [r8+1]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+2]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+3]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+4]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+5]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+6]
        add     r9d, edx
        cdq
        idiv    ebx
        lea     eax, [r8+7]
        add     r8d, 8
        add     r9d, edx
        cdq
        idiv    ebx
        add     r9d, edx
        cmp     r8d, 100000
        jne     .L3
        add     r9d, edi
        add     rsi, 4
        lea     r10, [rsp+40000]
        mov     DWORD PTR [rsi-4], r9d
        cmp     r10, rsi
        jne     .L2
        movsx   r11, edi
        xor     eax, eax
        mov     edi, OFFSET FLAT:.LC0
        mov     esi, DWORD PTR [rsp+r11*4]
        call    printf
        mov     rbx, QWORD PTR [rbp-8]
        xor     eax, eax
        leave
        ret
```

**Comment by nervenes** [comment]
> @nervenes Nothing I suggested makes it no longer compare equal work.

I know, i'm pointing out that the benchmarks *already* doesn't compare equal work.

> You state this is fact but it doesn't and it takes a few seconds on godbolt just to verify that with any C or C++ compiler: https://godbolt.org/z/j3GWehxed.

I stand corrected, I shouldn't have used the word "fact".

**Comment by rsb-23** [comment]
There are so many unnecessary issues/PR/discussions, just because of 1 reason.
There is no defined benchmarking rules mentioned by author @bddicken .

I suggest him to add `BENCHMARKING_RULES.md` for his project, else there is no point creating issues and PRs.

**Maybe**,
- Code should use only builtin libraries
- Language version should be latest stable version as on 1st Dec 2024.
- Each benchmark should be simple and test only 1 aspect.

**Comment by withinboredom** [comment]
Yeah, I was looking at the benchmarks and realized they don't compare anything well. For example, seeing PHP near dead last caught me by surprise.

Here's a PHP implementation that handily beats java:

```
function fibonacciIterative(int $n): int
{
	if ($n == 0) return 0;
	if ($n == 1) return 1;

	$a = 0; // F(0)
	$b = 1; // F(1)

	for ($i = 2; $i <= $n; $i++) {
		[$a, $b] = [$b, $a + $b];
	}

	return $b; // Return F(n)
}
```

What you're actually testing in this benchmark is how fast growing the stack is, not how fast the language is by any means. For something like Java, it will optimize all the recursive calls away (making the implementation closer to that implementation above); PHP should be beating nearly every other language—it is written in highly optimized C, after all. BUT if you have to keep growing the stack to enter functions, that is what you'll be measuring and where you see the performance impact.

> To actually benchmark the runtime (i.e code execution as you're referring it to) you'd need a long running process actually computing something

Exactly.

**Comment by bddicken** [comment]
The benchmark has evolved since this issue was first created. While there may be a number of good points brought up here, I'm going to close this out so we can clean up the issues. Feel free to open new one(s) if you feel strongly about any one of these points specifically.
